### ISSUE TITLE
Improve type clarity of twisted.conch.ssh.userauth

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8240]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8240|
|Type|enhancement|
|Created|2016-03-21 01:07:05Z|
|Branch|https://github.com/twisted/twisted/tree/clarity-sshuserauth-8240|

As with [#8237](https://github.com/twisted/twisted/issues/8237), make porting easier by improving clarity of types.

<details><summary>Searchable metadata</summary>

```
trac-id__8240 8240
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__None None
branch__branches_clarity_sshuserauth_8240 branches/clarity-sshuserauth-8240
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__None None
time__1458522425215453 1458522425215453
changetime__1458614583556530 1458614583556530
version__None None
owner__hawkowl hawkowl
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

(In [47035]) Branching to clarity-sshuserauth-8240.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

Okay, this one's a bit smaller.

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Thanks for the changes.
Happy to see that twisted.conch got some attention

1. interfaceToMethod changes. I think that this required a warning in case some is setting non bytes values.

For my project I am using custom checkers and they are defined as just `str`.

Maybe we can make this private and deprecate it, and add a helper for updating this structure. Maybe `SSHUserAuthServer.addCredentialsHandler(

Same for supportedAuthentications

I am talking about this differece in py2 vs py3

```

Python 3.4.3
>>> ca = {}
>>> ca['b'] = 'b'
>>> ca[b'b'] = b'b'
>>> ca
{'b': 'b', b'b': b'b'}


Python 2.7.6 
>>> ca = {}
>>> ca['b'] = 'b'
>>> ca[b'b'] = b'b'
>>> ca
{'b': 'b'}

```

---------

I would say that changing from str to bytes without any deprecation warning is bad.

You can revert the type changes for interfaceToMethod and supportedAuthentications and merge

Otherwise please add your comments about why this is not breaking backward compatibility and resubmit it for review.

Many thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl set status to `closed`|
|-|-|

(In [47044]) Merge clarity-sshuserauth-8240: Increase type clarity in twisted.conch.ssh.userauth

Author: hawkowl
Reviewer: adiroiban
Fixes: [#8240](https://github.com/twisted/twisted/issues/8240)
